src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -300,35 +318,44 @@ private IList<DocumentId> GetDocumentIdsFromDocCookie(uint docCookie)
         private void CloseDocuments(uint docCookie, string monikerToKeep)
         {
             List<DocumentKey> documentKeys;
-            if (_docCookiesToOpenDocumentKeys.TryGetValue(docCookie, out documentKeys))
+            if (!_docCookiesToOpenDocumentKeys.TryGetValue(docCookie, out documentKeys))
             {
-                // We will remove from documentKeys the things we successfully closed,
-                // so clone the list so we can mutate while enumerating
-                var documentsToClose = documentKeys.Where(key => !StringComparer.OrdinalIgnoreCase.Equals(key.Moniker, monikerToKeep)).ToList();
-
-                // For a given set of open linked or shared files, we may be closing one of the
-                // documents (e.g. excluding a linked file from one of its owning projects or
-                // unloading one of the head projects for a shared project) or the entire set of
-                // documents (e.g. closing the tab of a shared document). If the entire set of
-                // documents is closing, then we should avoid the process of updating the active
-                // context document between the closing of individual documents in the set. In the
-                // case of closing the tab of a shared document, this avoids updating the shared 
-                // item context hierarchy for the entire shared project to head project owning the
-                // last documentKey in this list.
-                var updateActiveContext = documentsToClose.Count == 1;
-
-                foreach (var documentKey in documentsToClose)
+                // let others know about non roslyn document close
+                var buffer = TryGetTextBufferFromDocData(RunningDocumentTable.GetDocumentData(docCookie));
+                if (buffer != null)
```

Comment:
What happens if the document, after already being open, was then added to the workspace? At this point this code path will get skipped, and the view unsubscribe might not run?

Trying to do this state tracking in this class really isn't the right thing to do. We should just have a separate class to subscribe to the text view notifications.